### DISCUSSION TITLE
Using dovi_tool to inject dolby metadata into PNG, TGA or other image formats

### DISCUSSION THREAD

**Discussion by docdude** [quietvoid/dovi_tool#48] [Q&A] [answered]
We have developed a pattern generator for the raspberry PI to aid in calibration of colorspace for most types of displays.  Currently we can calibrate SDR, HDR colorspaces using the pattern generator and calibration software.  HDR metadata is injected by setting and EGL surface for BT2020 colorspace and setting the HDR metadata to BT2020 chroma values through the DRM (Direct rendering manager) via the kernel.  
Is it possible to inject DV metadata into a PNG image, e.g. [https://github.com/sandflow/hdr4png](url) can inject iCCP HDR metadata into a PNG image, or to inject the DV metadata directly through the DRM as we do currently with HDR metadata and passthrough to HDMI to set a capable display to DV.   This would help in allowing us to calibrate the Dolby Vision colorspace and possibly HDR10+.
Thanks for your efforts.

**Comment by quietvoid** [comment, marked as answer]
I don't think it's possible in an image format, maybe AVIF because it can contain generic ITU T-35 metadata, which supports most ST2094 applications.

Other than that you would have to replicate the metadata embedding in the HDMI signal, because DRM only specifies HDR10 metadata support.
This patent has relevant information: https://patentimages.storage.googleapis.com/1d/96/ae/9a569883484ca7/WO2017079132A1.pdf

**Reply by docdude** [reply]
Thank you, will look into this format.  Figured the HDMI packet would have to be the target.

**Comment by docdude** [comment]
Hi
I was able to manage adding hvs infoframes to kernel and code to parse Dolby VSDB .  We are using it in our pattern generator to calibrate Dolby compliant displays.  However we use standard Dolby infoframes in 1080p60 with images and patterns contains a pixel sequence to activate the display once it’s received the correct infoframe.  We do this with raspberry pi. Our problem and that of most pattern generators cannot do Low latency DoVi. We can send the infoframe and pattern but the colorspace values are all over the place..  I’ve reverse engineered the shaders in libplacebo for decode but we were thinking maybe encoding a pattern into Dolby colorspace might be the work around. 
Could then feed the shader the different RPU metadata profiles.  Has Antoine worked out the encoding matrices?

Regards 
Juan

Darkness cannot drive out darkness; only light can do that. Hate cannot drive out hate; only love can do that.

Martin Luther King, Jr.

**Reply by quietvoid** [reply]
> Has Antoine worked out the encoding matrices?

I don't know who you're talking about.

Properly encoding with useful polynomials has not been reversed.
See here for generating patterns with static values: https://github.com/test-full-band/tfb-video

There is one test pattern provided by Portrait Displays here: https://app.spectracal.com/dnl/LG/1hr_gray_DV_Profile5.mp4
Might be useful.

It might be better to start off with regular HDR10 based profiles, though.
I don't know what colorspace the TV receives from source-based mapping (LLDV).

**Comment by docdude** [comment]
Lol, damn iPhone autocorrect, meant anyone not Antoine …. Lol

Darkness cannot drive out darkness; only light can do that. Hate cannot drive out hate; only love can do that.

Martin Luther King, Jr.

**Comment by docdude** [comment]
Yeah I’ve looked thru that code, it generates video, I’m looking for a way  to encode test patterns 

Darkness cannot drive out darkness; only light can do that. Hate cannot drive out hate; only love can do that.

Martin Luther King, Jr.

**Comment by docdude** [comment]
As I mentioned we have patterns that trigger lldv and standard dv. These are used by LG, Murideo and others for calibration, they are png images with embedded pixel sequences.  When u trigger the display into DoVi with the proper hsv infoframe, the image will render with these pixels. We can also send a pattern embedded with these pixels in the right position and with the right rgb values. Without the pixels the screen will be black in std DoVi and the wrong color in lldv.  Weare able to calibrate Dolby displays using std DoVi but have been  unsuccessful with lldv. Hence the idea to encode patterns.   



Darkness cannot drive out darkness; only light can do that. Hate cannot drive out hate; only love can do that.

Martin Luther King, Jr.
